scripting/practicemode/grenade_utils.sp

```diff
@@ -792,6 +884,64 @@ public Action CorrectGrenadeIdsHelper(const char[] ownerName, const char[] owner
   g_NewKv.Rewind();
 }
 
+// Rethrows all grenades, and uses the CSU managed explosion forward to save grenade data.
+public void CorrectGrenadeDetonations(int initiatingClient) {
+  g_ManagedGrenadeDetonationsToFixPhase = GRENADE_DETONATION_FIX_PHASE_SMOKES;
+  IterateGrenades(_CorrectGrenadeDetonations_Iterator, initiatingClient);
+  PM_Message(initiatingClient, "Fixing detonations in %i phases.", g_ManagedGrenadeDetonationsToFixPhase);
+}
+
+public int CorrectGrenadeDetonationsAdvanceToNextPhase(int initiatingClient) {
+  g_ManagedGrenadeDetonationsToFixPhase -= 1;
+  if (g_ManagedGrenadeDetonationsToFixPhase > GRENADE_DETONATION_FIX_PHASE_DONE) {
+    IterateGrenades(_CorrectGrenadeDetonations_Iterator, initiatingClient);
+  } else {
+    g_ManagedGrenadeDetonationsToFixPhase = GRENADE_DETONATION_FIX_PHASE_DONE;
+  }
+  return g_ManagedGrenadeDetonationsToFixPhase;
+}
+
+public Action _CorrectGrenadeDetonations_Iterator(
+  const char[] ownerName, 
+  const char[] ownerAuth, 
+  const char[] name,
+  const char[] description, 
+  ArrayList categories,
+  const char[] grenadeID, 
+  const float origin[3], 
+  const float angles[3], 
+  const char[] grenadeTypeStr, 
+  const float grenadeOrigin[3], 
+  const float grenadeVelocity[3], 
+  const float grenadeDetonationOrigin[3], 
+  any initiatingClient
+) {
+  GrenadeType grenadeType = GrenadeTypeFromString(grenadeTypeStr);
+  if (
+    grenadeType == GrenadeType_None
+    || (grenadeType == GrenadeType_Smoke && g_ManagedGrenadeDetonationsToFixPhase == GRENADE_DETONATION_FIX_PHASE_NONSMOKES)
+    || (grenadeType != GrenadeType_Smoke && g_ManagedGrenadeDetonationsToFixPhase == GRENADE_DETONATION_FIX_PHASE_SMOKES) 
+  ) {
+    return;
+  }
+
+  // TODO: throw in 3 phases to break glass on some maps?
```

Comment:
Haha, I've never figured out a good way to deal with glass, the todo is good with me.